### PR TITLE
[FEATURE] #97305 - Introduce CSRF-like request-token handling

### PR DESCRIPTION
**Pull request by brotkrueml**
Resolves: https://github.com/TYPO3-Documentation/Changelog-To-Doc/issues/192
Resolves: https://github.com/TYPO3-Documentation/Changelog-To-Doc/issues/296
Releases: main

### CODE REVIEW

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+================================
+CSRF-like request token handling
+================================
+
+..  versionadded:: 12.0
+
+A CSRF-like request token handling is available to mitigate potential cross-site
+requests on actions with side effects. This approach does not require an
+existing server-side user session, but uses a nonce (number used once) as a
+"pre-session". The main scope is to ensure a user actually has visited a page,
```
Comment (linawolf) [inline Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst:14]: I don't understand it, can we link somewhere or explain pre-sessions and nonce?

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+before submitting data to the webserver.
+
+This token can only be used for HTTP methods `POST`, `PUT` or `PATCH`, but
+for instance not for `GET` request.
+
+The :php:`\TYPO3\CMS\Core\Middleware\RequestTokenMiddleware` resolves
```
Comment (linawolf) [inline Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst:20]: link this to github?

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+request tokens and nonce values from a request and enhances responses with
+a nonce value in case the underlying application issues one. Both items are
+serialized as JSON Web Token (JWT) hash signed with `HS256`. Request tokens
```
Comment (linawolf) [inline Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst:23]: Do readers have to understand these technological deatails? If so I would link somewhere for additional information

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+use the provided nonce value during signing.
+
+Session cookie names involved for providing the nonce value:
+
+*   `typo3nonce_[hash]` in case request served with plain HTTP
+*   `__Secure-typo3nonce_[hash]` in case request served with secured HTTPS
+
+Submitting request token value to application:
+
+*   HTTP body, e.g. in `<form>` via parameter `__RequestToken`
```
Comment (linawolf) [inline Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst:33]: avoid abreviations

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+*   HTTP header, e.g. in XHR via header `X-TYPO3-Request-Token`
+
+..  attention::
+    When working with multiple browser tabs, an existing nonce value (stored as
+    session cookie in the browser of the user) might be overridden.
+
+..  note::
+    The current concept uses the :php:`\TYPO3\CMS\Core\Security\NoncePool` which
```
Comment (linawolf) [inline Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst:41]: link this to github?

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+The sequence looks like the following:
+
+..  rst-class:: bignums-xxl
+
+#.  Retrieve nonce and request token values
+
+    This happens on the previous legitimate visit on a page that offers
+    a corresponding form that shall be protected. The `RequestToken` and `Nonce`
+    objects (later created implicitly in this example) are organized in the
+    :php:`\TYPO3\CMS\Core\Context\SecurityAspect`.
```
Comment (linawolf) [inline Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst:57]: link this

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    extracted from markup and used without having the possibility to verify the
+    procedural integrity.
+
+    The middleware :php:`\TYPO3\CMS\Core\Middleware\RequestTokenMiddleware`
+    takes care of providing the received nonce and received request token values
+    in :php:`\TYPO3\CMS\Core\Context\SecurityAspect`. The handling controller
+    action needs to verify that the request token has the expected
+    `'my/process'` scope.
+
+    ..  code-block:: php
```
Comment (linawolf) [inline Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst:121]: move this to a file?

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst (hunk 1 above)
Comment (brotkrueml) [reply Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst:14]: I understand pre-session as something that must be valid (request token) prior login. So you have to visit the page before posting the credentials. That's why it is in quotes.

Comment (brotkrueml) [reply Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst]: Nonce is linked to the almighty Wikipedia :)

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst (hunk 2 above)
Comment (brotkrueml) [reply Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst:20]: Does it has any value? It is no API, it just does the work?

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst (hunk 3 above)
Comment (brotkrueml) [reply Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst]: I assume, if someone wants to create a custom authentication, they should know what they are doing and are familiar with these concepts. I consider this chapter as "for specialists".

Comment (brotkrueml) [reply Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst]: But JWT is linked.

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst (hunk 5 above)
Comment (brotkrueml) [reply Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst:41]: I see no value for the user (also this class is marked as internal). It is interesting to look into it from a "how is this done" view.

File: Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst (hunk 7 above)
Comment (brotkrueml) [reply Documentation/ApiOverview/Authentication/CSRFlikeRequestTokenHandling.rst:125]: Thought about it, but it is not really valid (the view is not initialized here). Can do that, of course, but then we have more distracting code from the substantial things.

**Review by linawolf — APPROVED** [review]

**Pull request merged**